lib/api/model/events.dart

```diff
@@ -74,6 +81,75 @@ class AlertWordsEvent extends Event {
   Map<String, dynamic> toJson() => _$AlertWordsEventToJson(this);
 }
 
+/// A Zulip event of type `user_settings` with op `update`.
+sealed class UserSettingsUpdateEvent extends Event {
+  @override
+  @JsonKey(includeToJson: true)
+  String get type => 'user_settings';
+
+  String get op => 'update';
+
+  UserSettingsUpdateEvent({
+    required super.id,
+  });
+
+  factory UserSettingsUpdateEvent.fromJson(Map<String, dynamic> json) {
+    switch (UserSettingName.fromRawString(json['property'] as String)) {
+      case UserSettingName.twentyFourHourTime:
+        return UserSettingsUpdateEventTwentyFourHourTime.fromJson(json);
+      case UserSettingName.displayEmojiReactionUsers:
+        return UserSettingsUpdateEventDisplayEmojiReactionUsers.fromJson(json);
+      case null:
+        return UserSettingsUpdateEventUnknown.fromJson(json);
+    }
+  }
+}
+
+@JsonSerializable(fieldRename: FieldRename.snake)
+class UserSettingsUpdateEventTwentyFourHourTime extends UserSettingsUpdateEvent {
```

Comment:
One advantage of the separate subclasses (cumbersome as they are) is that, with the base class `UserSettingsUpdateEvent` being sealed, we can ensure an exhaustive check in `PerAccountStore.handleEvent`:

```dart
    } else if (event is UserSettingsUpdateEvent) {
      assert(debugLog("server event: user_settings/update"));
      switch (event) {
        case UserSettingsUpdateEventTwentyFourHourTime():
          userSettings?.twentyFourHourTime = event.value;
        case UserSettingsUpdateEventDisplayEmojiReactionUsers():
          userSettings?.displayEmojiReactionUsers = event.value;
        case UserSettingsUpdateEventUnknown():
          // do nothing
      }
      notifyListeners();
```

which provides the analysis error mentioned in step 3 in the added model/events test:

<details>

```dart
  test('user_settings: all known settings have event handling', () {
    final missingEnumNames = UserSettingName.missingNames();
    check(
      missingEnumNames,
      because:
        'You have added these fields to [UserSettings]\n'
        'without handling the corresponding forms of the\n'
        'user_settings/update event in [PerAccountStore]:\n'
        '  $missingEnumNames\n'
        'To do that, please follow these steps:\n'
        '  (1) Add corresponding members to the [UserSettingName] enum.\n'
        '  (2) Then, when you see a Dart analysis error about\n'
        '      not exhaustively matching on `UserSettingName?`,\n'
        '      where we represent user_settings/update events,\n'
        '      please handle it by adding a new case\n'
        '      on the pattern of the other cases you see there.\n'
        '  (3) Then, when you see another Dart analysis error about\n'
        '      not handling the new event type in [PerAccountStore],\n'
        '      please do so, on the pattern of the other cases there.'
    ).isEmpty();
  });
```

</details>

We do have the `UserSettingName` enum, and we know it's possible to enforce an exhaustive check on one of those values. So maybe that's a good alternative way to enforce an exhaustive check in `PerAccountStore`, instead of with a sealed class. However, in a switch on a `UserSettingName`, ideally each case would know what type of setting value (`bool`, `int`, etc.) it's dealing with, automatically, without making `PerAccountStore` repeat that information and risk getting it wrong. Does that make sense? I don't yet see a good way to do that, though.

Reply:
Yeah, the exhaustive check is definitely good to achieve. I guess a couple of possible ways to do that would be:

* With per-value-type subclasses like UserSettingsUpdateEventBool, have also a separate enum per value type. Then once you're looking at a UserSettingsUpdateEventBool, you can switch exhaustively on its `name` field because it'll come from the enum of boolean settings.

* Use per-value-type subclasses like UserSettingsUpdateEventBool, and keep a single enum for all the setting names; but to do an exhaustive switch, switch first on the `name` field of the base class UserSettingsUpdateEvent. Then the cases look something like:
```dart
  case UserSettingName.twentyFourHourTime:
    userSettings?.twentyFourHourTime = (event as UserSettingsUpdateEventBool).value;
```

* Like the previous option, but skip the subclasses and just make the base class generic on the type of `value`, because it's not clear at that point that the subclasses are doing much for us. Then the cases look like:
```dart
  case UserSettingName.twentyFourHourTime:
    userSettings?.twentyFourHourTime = (event as UserSettingsUpdateEvent<bool>).value;
```

* Like the previous option, but don't make it generic and instead just give `value` a type of `Object?`, because it's not clear at that point that we're making any static use of the generic type parameter. Then the cases look like:
```dart
  case UserSettingName.twentyFourHourTime:
    userSettings?.twentyFourHourTime = event.value as bool;
```

I think probably that last option beats the second or third; I'm not sure either of them ultimately have any advantages over the last one. The first option is more incommensurable; but the per-value type enums sound like kind of a pain, so I'd lean toward the last option.

What the last three options have in common is that they give up on having a static check that `handleEvent` (or other consuming code) agrees with the API binding on what value type corresponds to which setting. But I think that's OK, because I think that's a relatively low-risk class of potential bug — once you're actively adding code in all the places you need to, it's relatively easy to be sure they agree on the type.

The higher-risk class of potential bug is the one that exhaustiveness prevents, because it's a lot easier to forget some update-needing spot entirely.

Reply:
Sure, I'm happy to do that last option.

> I think probably that last option beats the second or third; I'm not sure either of them ultimately have any advantages over the last one.

My first thought was that those other options _do_ have an advantage over the last one. I thought the last one, unlike the others, retreats from a "crunchy shell" ideal by allowing these events to flow into `PerAccountStore.handleEvent` without first type-guarding some part of them that we care about (`value`).

But that's not necessary: even if the class has `dynamic value`, we can still have its `fromJson` do that runtime type-guarding. (With an exhaustive check!) Probably that's what you had in mind; what do you think?

Reply:
Yeah, definitely agree we'll want the `fromJson` to check it's the right type, the same way as it usually would.